### PR TITLE
Remove analyze_branch and BranchInfo

### PR DESCRIPTION
**Pull request by elliottt**
We don't have overlap in behavior for branch instructions anymore, so we can remove `analyze_branch` and instead match on the `InstructionData` directly.

### CODE REVIEW

File: cranelift/frontend/src/ssa.rs [hunk 1]
```diff
@@ -622,7 +618,7 @@ impl SSABuilder {
                     jt = func.create_jump_table(copied);
                 }
 
-                // Redo the match from `analyze_branch` but this time capture mutable references
+                // Redo the match from, but this time capture mutable references
                 match &mut func.dfg.insts[branch] {
                     InstructionData::BranchTable {
                         destination, table, ..
```
Comment (elliottt) [inline cranelift/frontend/src/ssa.rs:624]: It's worth noting that once `br_table` uses `BlockCall`, we can replace the outer match with a mutable match, and remove this inner re-match.

File: cranelift/frontend/src/ssa.rs [hunk 2]
```diff
@@ -622,7 +618,7 @@ impl SSABuilder {
                     jt = func.create_jump_table(copied);
                 }
 
-                // Redo the match from `analyze_branch` but this time capture mutable references
+                // Redo the match from, but this time capture mutable references
```
Comment (jameysharp) [inline cranelift/frontend/src/ssa.rs:621]:
```suggestion
                // Redo the match from above, but this time capture mutable references
```

**Review by jameysharp — APPROVED** [review]
I wonder how many of these former uses of `analyze_branch` could be replaced with calls to `visit_block_succs`? I'm pretty sure at least the ones in `dominator_tree.rs` and `flowgraph.rs` could be.

File: cranelift/frontend/src/ssa.rs (hunk 2 above)
Comment (elliottt) [reply cranelift/frontend/src/ssa.rs:621]: Yeah, that does read a bit better :sweat_smile:

**Comment by elliottt** [comment]
> I wonder how many of these former uses of `analyze_branch` could be replaced with calls to `visit_block_succs`? I'm pretty sure at least the ones in `dominator_tree.rs` and `flowgraph.rs` could be.

That would be a nice refactoring!

**Comment by jameysharp** [comment]
Just so we're clear: I did push the "approve" button, with or without that hypothetical refactoring. :grin:

**Comment by bjorn3** [comment]
Exception handling may introduce a new instruction with semantics identical to br_table with respect to determining successors. analyze_branch would still be useful for that. In addition when a new kind of branch instruction is added, this PR causes there to no longer be a compile error or runtime error when a match is missed. Previously forgetting to add it to analyze_branch would trigger the !is_branch assertion and if a new BranchInfo variant was added you did get a compile error.

**Comment by bjorn3** [comment]
Maybe the visit_block_succs refactoring could be done and the remaining cases could get the !is_branch assertion added back?

**Comment by elliottt** [comment]
> Exception handling may introduce a new instruction with semantics identical to br_table with respect to determining successors. analyze_branch would still be useful for that.

Would that have turned into a return value of `BranchInfo::Table` in `analyze_branch`? If so, you would still have needed to re-match the opcode to apply special behavior for your new instruction.

> In addition when a new kind of branch instruction is added, this PR causes there to no longer be a compile error or runtime error when a match is missed. Previously forgetting to add it to analyze_branch would trigger the !is_branch assertion and if a new BranchInfo variant was added you did get a compile error.

This is true, however my hope was that we've now arrived at the necessary set of branch instructions and won't need to add any others. We have unconditional jump, conditional branch, and branch through table instructions, which feel like the right building blocks for more complicated control flow. I think the `switch` module in cranelift-frontend is a great example of this.

**Comment by bjorn3** [comment]
> Would that have turned into a return value of BranchInfo::Table in analyze_branch?

Yes

> If so, you would still have needed to re-match the opcode to apply special behavior for your new instruction.

Yeah, in some cases that is indeed necessary. In other cases it doesn't actually care about which exact branch instruction is used. Just which successors exist.

> This is true, however my hope was that we've now arrived at the necessary set of branch instructions and won't need to add any others. We have unconditional jump, conditional branch, and branch through table instructions, which feel like the right building blocks for more complicated control flow. I think the switch module in cranelift-frontend is a great example of this.

I see. For exceptions I am pretty sure we will need an invoke/invoke_indirect instruction to be able to define the unwinding path that a call will take. From the clif ir perspective this will be a branch instruction which branches when an exception happens. I don't think this is implementable without support from cranelift-codegen.

**Comment by cfallin** [comment]
+1 to `invoke`-family instructions: fundamentally we can't represent the zero-cost exception model without it, I think, because the unwinder is external and effectively jumps to a different return address.

**Comment by elliottt** [comment]
> I see. For exceptions I am pretty sure we will need an invoke/invoke_indirect instruction to be able to define the unwinding path that a call will take. From the clif ir perspective this will be a branch instruction which branches when an exception happens. I don't think this is implementable without support from cranelift-codegen.

I'll make a PR that both implements @jameysharp's earlier refactoring suggestion, and add `assert!(!inst.is_branch())` to all the cases that I modified when removing `BranchInfo`. Does that seem sufficient for not making the addition of `invoke`/`invoke_indirect` not too painful?

**Comment by bjorn3** [comment]
I think that is sufficient.

**Pull request merged**